### PR TITLE
Migrate GitHub to RetroFit , add pagination and token support

### LINKED ISSUE #10: Authentication with GitHub
Integration is in place, but authentication would be nice also

### LINKED ISSUE #15: Github, support pagination
Currently only 30 issues can be found.

### PR DESCRIPTION
**Pull request by spacecowboy**
Hi, great library and collection of plugins! I really want to start using it immediately but need token/pagination support for GitHub. So this PR fixes #10 and fixes #15. 

It is rather intrusive in that it also makes use of [RetroFit](http://square.github.io/retrofit/). It's a great library, but I understand if you're hesitant to accept such a large change. But on the plus side, I made sure to port the tests over as well.

I also updated the command-line version to accept a new argument: `-gtok/--github-token`, which I can issue as a PR as well.

Again, great library!

### CODE REVIEW

**Comment by tomasbjerre** [comment]
This is looking great! And thanks for showing me Retrofit  =)

I will test this later today and then release it!

File: src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java [hunk 1]
```diff
@@ -303,7 +313,11 @@ public void toMediaWiki(String username, String password, String url, String tit
   * Get the changelog as data object.
   */
  public Changelog getChangelog() {
-  return getChangelog(new GitRepo(new File(settings.getFromRepo())));
+  try {
+   return getChangelog(new GitRepo(new File(settings.getFromRepo())));
+  } catch (IOException e) {
+   return null;
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/api/GitChangelogApi.java]: Throwables.propagate(e)

File: src/main/java/se/bjurr/gitchangelog/internal/integrations/github/GitHubHelper.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    for (GitHubIssue gitHubIssue: response.body()) {
+     if (issue.equals(gitHubIssue.getNumber())) {
+      // Done
+      return of(gitHubIssue);
+     }
+    }
+
+   } catch (IOException e) {
+    System.out.println(e.getMessage());
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/integrations/github/GitHubHelper.java]: log

File: src/main/java/se/bjurr/gitchangelog/internal/integrations/github/GitHubServiceFactory.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+ public static void reset() {
+  gitHubService = null;
+ }
+
+ public static GitHubService getGitHubService(String api, final Optional<String> token) {
+  return getGitHubService(api, token, null);
+ }
+
+ public static GitHubService getGitHubService(String api, final Optional<String> token, final Interceptor interceptor) {
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/integrations/github/GitHubServiceFactory.java]: synchronized

File: src/main/java/se/bjurr/gitchangelog/internal/model/Transformer.java [hunk 4]
```diff
@@ -47,7 +49,12 @@ public Tag apply(GitTag input) {
     List<GitCommit> gitCommits = input.getGitCommits();
     List<Commit> commits = toCommits(gitCommits);
     List<Author> authors = toAuthors(gitCommits);
-    List<ParsedIssue> parsedIssues = new IssueParser(settings, gitCommits).parseForIssues();
+    List<ParsedIssue> parsedIssues;
+    try {
+     parsedIssues = new IssueParser(settings, gitCommits).parseForIssues();
+    } catch (IOException e) {
+     parsedIssues = new ArrayList<ParsedIssue>();
```
Comment (tomasbjerre) [inline src/main/java/se/bjurr/gitchangelog/internal/model/Transformer.java]: log, or throw?

Comment (spacecowboy) [inline src/main/java/se/bjurr/gitchangelog/internal/model/Transformer.java]: I'm all in favor of throwing. I just didn't want to change your API.

**Comment by tomasbjerre** [comment]
This is now released in 1.33.
I made some adjustments here: https://github.com/tomasbjerre/git-changelog-lib/commit/4cb3f757f8a6b9f9c8aa6dc7e88618fa90d6841f

Thanks again for this one!

**Comment by tomasbjerre** [comment]
Did you have a reason for choosing okhttp 2.3.0? When using the lib with the Maven plugin I got an error like below. Solved by using okhttp 2.7.5

```
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:166)
	... 21 more
Caused by: java.lang.NoClassDefFoundError: okio/ForwardingTimeout
	at okhttp3.internal.http.Http1xStream.newChunkedSource(Http1xStream.java:242)
	at okhttp3.internal.http.Http1xStream.getTransferStream(Http1xStream.java:139)
	at okhttp3.internal.http.Http1xStream.openResponseBody(Http1xStream.java:129)
	at okhttp3.internal.http.HttpEngine.readNetworkResponse(HttpEngine.java:732)
	at okhttp3.internal.http.HttpEngine.access$200(HttpEngine.java:81)
	at okhttp3.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:708)
	at okhttp3.internal.http.HttpEngine.readResponse(HttpEngine.java:563)
	at okhttp3.RealCall.getResponse(RealCall.java:241)
	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:198)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:160)
	at okhttp3.RealCall.execute(RealCall.java:57)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:174)
	at se.bjurr.gitchangelog.internal.integrations.github.GitHubHelper.getIssueFromAll(GitHubHelper.java:37)
	at se.bjurr.gitchangelog.internal.issues.IssueParser.parseForIssues(IssueParser.java:76)
	at se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:361)
	at se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:328)
	at se.bjurr.gitchangelog.api.GitChangelogApi.render(GitChangelogApi.java:344)
	at se.bjurr.gitchangelog.api.GitChangelogApi.toFile(GitChangelogApi.java:306)
	at se.bjurr.gitchangelog.plugin.GitChangelogMojo.execute(GitChangelogMojo.java:136)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
Caused by: java.lang.ClassNotFoundException: okio.ForwardingTimeout
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	... 41 more
```

**Comment by spacecowboy** [comment]
No reason at all, simply what came along with the _copy-paste_ :)

**Comment by tomasbjerre** [comment]
Ok :) Its in the command line tool, gradle and maven plugins now. I'll add it to the Jenkins plugin tomorrow.

**Pull request merged**